### ISSUE TITLE
Advice about using scope and spawn inside an object

### ISSUE THREAD

**Issue by LouChiSoft** [rayon-rs/rayon#973]
Hi, I realise this might be more of a Rust than a Rayon question so apologies if it is as I am still quite new to Rayon. I am trying to write a game engine in Rust and found Rayon as a threading library which looks like it will work wonders for looping over a vector of game objects and calling their respective `update()` function. But I wanted to also give the user of the engine the ability to create and submit tasks from within an objects `update()` function by creating other methods and then using a combination of `scope` and `Scope::spawn`s to submit the functions.

The issue is that `update()` functions will almost always change `self` in someway and trying to submit multiple tasks as functions that update self breaks the borrow checker rules of only 1 mutable borrow is allowed at a time. I can get around this if I use closures to capture the individual parts of the object to update and that runs fine, but trying to convert that into something that easy for the user to implement is proving a bit difficult and I wondered if anyone on the rayon team has any experience with a scenario similar to what I am describing.

Example code:
```rs
pub struct GameObject {
    position: (f32, f32, f32),
    rotation: (f32, f32, f32),
    scale:    (f32, f32, f32)
}

impl GameObject {

    pub fn new() -> GameObject {
        GameObject {
            position: (0_f32,0_f32,0_f32),
            rotation: (0_f32,0_f32,0_f32),
            scale:    (1_f32,1_f32,1_f32)
        }
    }

    // Working update function
    pub fn update(&mut self) {
    
        let scope = rayon::scope(|s| {
    
            // update position
            s.spawn(|_| {
                self.position = (1_f32, 1_f32, 1_f32);
            });
    
            // update rotation
            s.spawn(|_| {
                self.rotation = (270_f32, 180_f32, 90_f32);
            });
    
            // update scale
            s.spawn(|_| {
                self.scale = (2_f32, 1_f32, 1_f32);
            });
    
        });
    }

    // Borrow checker breaking update function
    pub fn update(&mut self) {

        let scope = rayon::scope(|s| {

            // update position
            s.spawn(|_| {
                self.update_position(1_f32, 1_f32, 1_f32);
            });

            // update rotation
            s.spawn(|_| {
                self.update_rotation(270_f32, 180_f32, 90_f32);
            });

            // update scale
            s.spawn(|_| {
                self.update_scale(2_f32, 1_f32, 1_f32);
            });

        });
    }

    fn update_position(&mut self, x: f32, y: f32, z: f32) {
        self.position = (x,y,z);
    }

    fn update_rotation(&mut self, x: f32, y: f32, z: f32) {
        self.rotation = (x,y,z);
    }

    fn update_scale(&mut self, x: f32, y: f32, z: f32) {
        self.scale = (x,y,z);
    }

}
```

Ideally I would like the user to only need to write a simpler update function along the lines of:
```
pub fn update(&mut self) {
    let task_a = ScheduleTask!(self.update_position, 1_f32, 1_f32, 1_f32);
    let task_b = ScheduleTask!(self.update_rotation, 270_f32, 180_f32, 90_f32);
    let task_c = ScheduleTask!(self.update_scale,    2_f32, 1_f32, 1_f32);

    SubmitTasks!(task_a, task_b, task_c);
}
```
And the macros would behind the scene convert that into a working series of scoped spawns, but I am unsure of how to get that working. Even with macros I don't know if it would be possible to save the function invocations as variables to be able to inject into the SubmitTasks! macro later. Does anyone have experience with a similar scenario or know how this coud be handled in Rust?

**Comment by cuviper** [comment]
Right, that's a general Rust issue that `&mut self` can't be borrowed in multiple places at once. Your first version works because of the 2021 edition feature [disjoint capture in closures](https://doc.rust-lang.org/edition-guide/rust-2021/disjoint-capture-in-closures.html), so each spawn closure is only borrowing the directly-mentioned field. But anything that goes through a full `&mut self` method won't get that benefit.

On the Rayon side of things -- do you have a lot of game objects to process? If possible, I'd suggest coarser parallelism at that level, and let them individually update themselves without further rayon calls.

For a game engine, you might also want to look at what Bevy does. They used to use rayon, but they found it to be too aggressive about trying to find parallel work, especially for frame-based intervals (like a game) where work keeps coming in bursts. They announced their own task system [here](https://bevyengine.org/news/bevy-0-2/#async-task-system).

**Comment by LouChiSoft** [comment]
I didn't realise that the dijointed captures were a recent edition change that's interesting to know. In terms of the game objects to process my intent was that in the background every update cycle would be a vector of game objects that uses a `par_iter_mut()` to call the `update()` function of each game object. 
This would be fine for the most part but in the event that you have 1 object that is particularly CPU heavy (a game manager class, generator of some kind or the like) where it's not really feasible to have them as separate objects but in theory you could split the update into multiple segments I wanted that abililty to split the update into separate spawns.
Like I did above with the closure version of the update function it can be done with disjoint captures, but it requires not an insignificant amount of boiler plate code each time you want to submit new tasks and it's my personal belief that when it comes to scripting for gameplay or something similar the API should be as simple as possible, hence my desire to replace it all with `ScheduleTask!` and `SubmitTask!` calls.

Bevy swapping to their own version is interesting as I was originally going to roll my own version similar to one I had created in C++ a couple of years ago or so to start with. But due to language limitations of Rust (namely not supporting variadics in anything other than Macros) it just wasn't feasible to get working in the same way I had the C++ version. I have since learned of a way I could get most of the functionality of what I had in C++ but by the point I learned about it I had come across Rayon and was trying to integrate that. I might have to re-evaluate whether or not I roll my own which is a shame because I think the simplicity of Rayon's approach is quite genius, but in this case I value simplicity for the end user higher than simplicity for me to implement it.
